### PR TITLE
BZ-1373412 - [GSS](6.4.z) TimeoutException when request ends up on …

### PR DESCRIPTION
**Pull request by JiriOndrusek**
…non-coordinator node (invalidation cache, web session passivation and shared cache store)

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1373412
No upstream required

### CODE REVIEW

File: core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java [hunk 1]
```diff
@@ -85,21 +85,25 @@ public void init(CommandsFactory factory) {
...
             lockKeyAndCheckOwnership(ctx, command.getKey(), lockTimeout, skipLocking);
          }
          return invokeNextInterceptor(ctx, command);
       } catch (Throwable t) {
+         releaseRemote = false;
          releaseLocksOnFailureBeforePrepare(ctx);
          throw t;
       } finally {
          if (!ctx.isInTxScope()) lockManager.unlockAll(ctx);
+         if(releaseRemote) releaseRemoteIfNeeded(ctx, command, cdl.localNodeIsPrimaryOwner(command.getKey()));
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java]: This seems wrong, the locks should only be released on commit/rollback. If we release the lock immediately after reading it, then we might as well not lock it at all, like we do when the `FORCE_WRITE_LOCK` flag isn't used.

Comment (tristantarrant) [inline core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java]: @JiriOndrusek can you address @danberindei 's comment ?

Comment (JiriOndrusek) [inline core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java]: @danberindei I see my mistake, I'll look for the correct solution.

**Comment by JiriOndrusek** [comment]
New fix is prepared.
@danberindei Could you look into it?

File: core/src/test/java/org/infinispan/interceptors/InvalidationInterceptorMockTest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+        RpcManager rpcManager = mock(RpcManager.class);
+        when(rpcManager.getTransport()).thenReturn(mock(Transport.class));
+
+        invalidationInterceptor.inject(rpcManager);
+        invalidationInterceptor.injectDependencies(commandsFactory);
+
+        invalidationInterceptor.invalidateAcrossCluster(true, null, ctx);
+
+        if(inTransaction) {
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/interceptors/InvalidationInterceptorMockTest.java]: Please add a space after `if`.

**Comment by danberindei** [comment]
@JiriOndrusek You're missing the `EntryWrappingVisitor.visitInvalidateCommand()` addition from https://github.com/infinispan/infinispan/pull/1726/files. This causes `SyncInvalidationTest` to fail, because the `InvalidationCommand` wrapped in a `PrepareCommand` is not processed and the value is not removed.

Also, please remove the `[]` around `BZ-1373412` from the commit message.

**Comment by JiriOndrusek** [comment]
@danberindei all changes you suggested are commited into PR

**Comment by danberindei** [comment]
Integrated, thanks Jiri!

**Pull request closed without merging**